File: 120.Triangle.go
// https://leetcode.com/problems/triangle/description/
//
// algorithms
// Medium (36.97%)
// Total Accepted:    154.4k
// Total Submissions: 417.6k
// beats 100.0% of golang submissions

package leetcode

import "sort"

func minimumTotal(triangle [][]int) int {
	length := len(triangle)
	if length == 0 {
		return 0
	}

	arrSize := len(triangle[length-1])
	pathArr := make([]int, arrSize)

	for i := 0; i < len(triangle[0]); i++ {
		pathArr[i] = triangle[0][i]
	}

	level := 1
	for {
		if level == length {
			break
		}
		temp := make([]int, arrSize)
		temp[0] = pathArr[0] + triangle[level][0]
		for j := 1; j < level; j++ {
			temp[j] = minInt(pathArr[j], pathArr[j-1]) + triangle[level][j]
		}
		temp[level] = pathArr[level-1] + triangle[level][level]
		level++
		copy(pathArr, temp)
	}

	sort.Sort(sort.IntSlice(pathArr))

	return pathArr[0]
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 120.Triangle.java
// https://leetcode.com/problems/triangle/
//
// algorithms
// Medium (39.28%)
// Total Accepted: 180,880
// Total Submissions: 460,446

class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int len = triangle.size();
        if (len == 0) {
            return 0;
        }
        if (len == 1) {
            return triangle.get(0).get(0);
        }

        for (int i = 1; i < len; i++) {
            List<Integer> l = triangle.get(i);
            for (int j = 0; j <= i; j++) {
                int value = -1;
                if (j == 0) {
                    value = l.get(j) + triangle.get(i - 1).get(0);
                } else if (j == i) {
                    value = l.get(j) + triangle.get(i - 1).get(j - 1);
                } else {
                    value = l.get(j) + Math.min(triangle.get(i - 1).get(j - 1), triangle.get(i - 1).get(j));
                }
                l.set(j, value);
            }
        }

        int res = Integer.MAX_VALUE;

        for (int n : triangle.get(len - 1)) {
            res = Math.min(res, n);
        }

        return res;
    }
}